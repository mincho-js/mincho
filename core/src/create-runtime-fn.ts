import type {
  PatternResult,
  RuntimeFn,
  VariantGroups,
  VariantSelection,
} from './types';

const shouldApplyCompound = <Variants extends VariantGroups>(
  compoundCheck: VariantSelection<Variants>,
  selections: VariantSelection<Variants>,
  defaultVariants: VariantSelection<Variants>
) => {
  for (const key of Object.keys(compoundCheck)) {
    if (compoundCheck[key] !== (selections[key] ?? defaultVariants[key])) {
      return false;
    }
  }

  return true;
};

export const createRuntimeFn = <Variants extends VariantGroups>(
  config: PatternResult<Variants>
) => {
  const runtimeFn: RuntimeFn<Variants> & {
    macaronMeta: {
      variants: Array<keyof Variants>;
      defaultClassName: string;
      variantConcat: (variants: VariantSelection<Variants>) => string;
    };
  } = options => {
    let className = config.defaultClassName;

    const selections: VariantSelection<Variants> = {
      ...config.defaultVariants,
      ...options,
    };
    for (const variantName in selections) {
      const variantSelection =
        selections[variantName] ?? config.defaultVariants[variantName];

      if (variantSelection != null) {
        let selection = variantSelection;

        if (typeof selection === 'boolean') {
          // @ts-expect-error
          selection = selection === true ? 'true' : 'false';
        }

        const selectionClassName =
          // @ts-expect-error
          config.variantClassNames[variantName]?.[selection];

        if (selectionClassName) {
          className += ' ' + selectionClassName;
        }
      }
    }

    for (const [compoundCheck, compoundClassName] of config.compoundVariants) {
      if (
        shouldApplyCompound(compoundCheck, selections, config.defaultVariants)
      ) {
        className += ' ' + compoundClassName;
      }
    }

    return className;
  };

  runtimeFn.macaronMeta = {
    variants: Object.keys(config.variantClassNames),
    defaultClassName: config.defaultClassName,
    variantConcat(options) {
      let className = config.defaultClassName;

      for (const variantName in options) {
        const variantSelection = options[variantName];

        if (variantSelection != null) {
          let selection = variantSelection;

          if (typeof selection === 'boolean') {
            // @ts-expect-error
            selection = selection === true ? 'true' : 'false';
          }

          const selectionClassName =
            // @ts-expect-error
            config.variantClassNames[variantName]?.[selection];

          if (selectionClassName) {
            className += ' ' + selectionClassName;
          }
        }
      }

      return className;
    },
  };

  return runtimeFn;
};
