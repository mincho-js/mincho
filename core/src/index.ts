export * from '@vanilla-extract/recipes';
export * from '@vanilla-extract/css';
