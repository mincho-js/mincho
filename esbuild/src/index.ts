import { processVanillaFile } from '@vanilla-extract/integration';
import { Plugin as EsbuildPlugin } from 'esbuild';
import { dirname, join } from 'path';
import { babelTransform, compile } from '@macaron-css/integration';

/*
  -> load /(t|j)sx?/
  -> extract css and inject imports (extracted_HASH.css.ts)
    -> resolve all imports
    -> load imports, bundle code again
      -> add file scope to all files
      -> evaluate and generate .vanilla.css.ts files
      -> gets resolved by @vanilla-extract/esbuild-plugin
    -> process the file with vanilla-extract
    -> resolve with js loader
*/

interface MacaronEsbuildPluginOptions {
  includeNodeModulesPattern?: RegExp;
}

export function macaronEsbuildPlugin({
  includeNodeModulesPattern,
}: MacaronEsbuildPluginOptions = {}): EsbuildPlugin {
  return {
    name: 'macaron-css-esbuild',
    setup(build) {
      let resolvers = new Map<string, string>();
      let resolverCache = new Map<string, string>();

      build.onEnd(() => {
        resolvers.clear();
        resolverCache.clear();
      });

      build.onResolve({ filter: /^extracted_(.*)\.css\.ts$/ }, async args => {
        if (!resolvers.has(args.path)) {
          return;
        }

        let resolvedPath = join(args.importer, '..', args.path);

        return {
          namespace: 'extracted-css',
          path: resolvedPath,
          pluginData: {
            path: args.path,
            mainFilePath: args.pluginData.mainFilePath,
          },
        };
      });

      build.onLoad(
        { filter: /.*/, namespace: 'extracted-css' },
        async ({ path, pluginData }) => {
          const resolverContents = resolvers.get(pluginData.path)!;
          const { source, watchFiles } = await compile({
            esbuild: build.esbuild,
            filePath: path,
            originalPath: pluginData.mainFilePath!,
            contents: resolverContents,
            externals: [],
            cwd: build.initialOptions.absWorkingDir,
            resolverCache,
          });

          try {
            const contents = await processVanillaFile({
              source,
              filePath: path,
              outputCss: undefined,
              identOption:
                undefined ?? (build.initialOptions.minify ? 'short' : 'debug'),
            });

            return {
              contents,
              loader: 'js',
              resolveDir: dirname(path),
            };
          } catch (error) {
            if (error instanceof ReferenceError) {
              return {
                errors: [
                  {
                    text: error.toString(),
                    detail:
                      'This usually happens if you use a browser api at the top level of a file being imported.',
                  },
                ],
              };
            }

            throw error;
          }
        }
      );

      build.onLoad({ filter: /\.(j|t)sx?$/ }, async args => {
        if (args.path.includes('node_modules')) {
          if(!includeNodeModulesPattern) return;
          if(!includeNodeModulesPattern.test(args.path)) return;
        };

        // gets handled by @vanilla-extract/esbuild-plugin
        if (args.path.endsWith('.css.ts')) return;

        const {
          code,
          result: [file, cssExtract],
        } = await babelTransform(args.path);

        // the extracted code and original are the same -> no css extracted
        if (file && cssExtract && cssExtract !== code) {
          resolvers.set(file, cssExtract);
          resolverCache.delete(args.path);
        }

        return {
          contents: code!,
          loader: args.path.match(/\.(ts|tsx)$/i) ? 'ts' : 'js',
          pluginData: {
            mainFilePath: args.path,
          },
        };
      });
    },
  };
}

export const macaronEsbuildPlugins = () => [
  macaronEsbuildPlugin(),
  require('@vanilla-extract/esbuild-plugin').vanillaExtractPlugin(),
];
