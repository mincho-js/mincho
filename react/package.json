{
  "name": "@macaron-css/react",
  "version": "1.5.1",
  "license": "MIT",
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "repository": {
    "type": "git",
    "url": "https://github.com/mokshit06/macaron.git",
    "directory": "packages/react"
  },
  "exports": {
    ".": {
      "import": "./dist/index.mjs",
      "require": "./dist/index.js"
    },
    "./dist/*": "./dist/*",
    "./runtime": {
      "import": "./dist/runtime.mjs",
      "require": "./dist/runtime.js"
    }
  },
  "dependencies": {
    "@macaron-css/core": "1.5.1"
  },
  "files": [
    "dist",
    "src"
  ],
  "peerDependencies": {
    "@vanilla-extract/recipes": "^0.2.5",
    "react": ">=17",
    "react-dom": ">=17"
  },
  "devDependencies": {
    "@types/react": "^18.0.12",
    "@types/react-dom": "^18.0.6",
    "@vanilla-extract/recipes": "^0.2.5",
    "react": "^18.1.0",
    "react-dom": "^18.2.0"
  }
}
