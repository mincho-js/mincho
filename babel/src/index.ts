import { template, types as t, PluginObj, NodePath } from '@babel/core';
import generate from '@babel/generator';
import hash from '@emotion/hash';
import { PluginState, PluginOptions } from './types';
import { buildImport, buildModuleExport, buildNamedExport } from './utils';
import { styledComponentsVisitor } from './visitors/styled-components';

export { PluginOptions };

// TODO: target VariableDeclarator instead of VariableDeclaration
export function macaronBabelPlugin(
  opts: PluginOptions
): PluginObj<PluginState> {
  return {
    name: 'macaron-css-babel',
    visitor: {
      Program: {
        enter(path, state) {
          state.dependentNodes = new Set();
          state.styledNodes = [];

          path.traverse(styledComponentsVisitor(path), state);
        },
        exit(path, state) {
          const ast = buildModuleExport({
            nodes: t.objectExpression(
              state.styledNodes.map(node =>
                t.objectProperty(
                  t.identifier(node.ident),
                  t.identifier(node.ident)
                )
              )
            ),
          });
          const program = t.program(
            [...(state.dependentNodes as any), ast].map(path => {
              if (!('parent' in path)) return path;
              if (path.parent.type === 'Program') {
                return path.node as any;
              } else {
                return path.parent as any;
              }
            })
          );
          const cssExtract = generate(program).code;
          const generatedCode = generate(path.node).code;

          const cssFile = `extracted_${hash(generatedCode)}.css.ts`;

          // TODO: convert multiple imports/exports to one
          for (const name of state.styledNodes) {
            path.unshiftContainer(
              'body',
              buildImport({
                specifier: t.identifier(name.ident),
                alias: t.identifier(name.ident),
                source: t.stringLiteral(cssFile),
              })
            );
            if (name.shouldExport) {
              path.pushContainer(
                'body',
                buildNamedExport({ ident: name.ident })
              );
            }
          }

          // this is commented out because
          // it was causing issue with other babel plugins
          // like babel-preset-solid
          // which depends on scope state to add tmpl nodes

          // path.scope.crawl();

          opts.result = [cssFile, cssExtract];
        },
      },
      VariableDeclaration(variablePath, variableState) {
        const declarations = variablePath.get('declarations');

        if (declarations.length < 1) return;

        const decl = declarations[0].get('init');
        const id = declarations[0].get('id');

        if (!decl.isCallExpression()) return;
        if (!id.isArrayPattern() && !id.isIdentifier()) return;

        const extractionAPIs = [
          // @vanilla-extract/css
          'style',
          'styleVariants',
          'globalStyle',
          'createTheme',
          'createGlobalTheme',
          'createThemeContract',
          'createGlobalThemeContract',
          'assignVars',
          'createVar',
          'fallbackVar',
          'fontFace',
          'globalFontFace',
          'keyframes',
          'globalKeyframes',
          'style',
          'styleVariants',
          'globalStyle',
          'createTheme',
          'createGlobalTheme',
          'createThemeContract',
          'createGlobalThemeContract',
          'assignVars',
          'createVar',
          'fallbackVar',
          'fontFace',
          'globalFontFace',
          'keyframes',
          'globalKeyframes',
          // @vanilla-extract/recipes
          'recipe',
        ];

        if (
          !extractionAPIs.some(api =>
            decl.get('callee').referencesImport('@macaron-css/core', api)
          )
        ) {
          return;
        }

        for (const key in decl.scope.bindings) {
          const binding = decl.scope.bindings[key];
          if (binding && key) {
            variableState.dependentNodes.add(binding.path);
          }
        }

        const shouldExport = variablePath.parentPath.isExportNamedDeclaration();

        if (id.isArrayPattern()) {
          for (const elementPath of id.get('elements')) {
            if (elementPath.isIdentifier()) {
              variableState.styledNodes.push({
                shouldExport,
                ident: elementPath.node.name,
              });
            }
          }
        } else {
          variableState.styledNodes.push({
            shouldExport,
            ident: id.node.name,
          });
        }

        variablePath.remove();
        variablePath.scope.crawl();
      },
    },
  };
}
