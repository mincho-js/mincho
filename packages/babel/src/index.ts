import { PluginObj, transformSync } from "@babel/core";
import { transformCallExpression } from "@/transforms/callExpression.js";
import postprocess from "@/transforms/postprocess.js";
import preprocess from "@/transforms/preprocess.js";
import { PluginOptions, PluginState } from "@/types.js";
import { styledComponentPlugin } from "@/styled.js";

export function minchoBabelPlugin(): PluginObj<PluginState> {
  return {
    name: "mincho-babel-plugin",
    visitor: {
      Program: {
        enter: preprocess,
        exit: postprocess
      },
      CallExpression: transformCallExpression
    }
  };
}

export { styledComponentPlugin as minchoStyledComponentPlugin } from "@/styled.js";
export type { PluginOptions } from "@/types.js";

// == Tests ====================================================================
// Ignore errors when compiling to CommonJS.
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore error TS1343: The 'import.meta' meta-property is only allowed when the '--module' option is 'es2020', 'es2022', 'esnext', 'system', 'node16', or 'nodenext'.
if (import.meta.vitest) {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore error TS1343: The 'import.meta' meta-property is only allowed when the '--module' option is 'es2020', 'es2022', 'esnext', 'system', 'node16', or 'nodenext'.
  const { describe, it, expect } = import.meta.vitest;

  function babelTransform(code: string) {
    const options: PluginOptions = { result: ["", ""] };
    const result = transformSync(code, {
      plugins: [[minchoBabelPlugin(), options], [styledComponentPlugin()]],
      presets: ["@babel/preset-typescript"],
      filename: "test.tsx"
    });

    if (result === null || result.code === null) {
      throw new Error("Failed to transform code");
    }

    return { result: options.result, code: result.code };
  }

  describe.concurrent("minchoBabelPlugin", () => {
    it("export default style", () => {
      const { result, code } = babelTransform(`
        import { style } from "@mincho-js/css";

        export default style({
          color: "red",
        });
      `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    it("inside jsx expression", () => {
      const { result, code } = babelTransform(`
        import { style } from '@mincho-js/css';

        function App() {
          return <div class={style({
            color: 'red'
          })}>Hello</div>
        }

        console.log(red);
      `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    it("hoists inline expression", () => {
      const { result, code } = babelTransform(`
        import { style } from '@mincho-js/css';
        const str = \`abc \${style({ color: "red" })}\`;
        console.log(str);
      `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    it("hoists object property", () => {
      const { result, code } = babelTransform(`
        import { style } from '@mincho-js/css';
        const obj = {
          nested: {
            key: style({ color: "red" })
          }
        };
        console.log(obj);
      `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    it("hoists array member", () => {
      const { result, code } = babelTransform(`
        import { style } from '@mincho-js/css';
        const arr = [1, 2, style({ color: "red" }), 4, style({ color: "blue" })];
        console.log(arr);
      `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    it("extracts style function", () => {
      const { result, code } = babelTransform(`
        import { style } from '@mincho-js/css';
        const red = style({ color: "red" });
        console.log(red);
      `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    it("extracts $mincho function", () => {
      const { result, code } = babelTransform(`
        import { style, mincho$ } from '@mincho-js/css';
        const red = mincho$(() => {
          return 2 + 2;
        });
        console.log(red);
      `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    it("multiple variable declarators in one declaration", () => {
      const { result, code } = babelTransform(`
      import { style } from '@mincho-js/css';

      const red = style({ color: 'red' }),
        blue = style({ color: 'blue' }),
        green = style({ color: 'green' });

      console.log(red, blue, green);
    `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    it("move bindings along with extracted style", () => {
      const { result, code } = babelTransform(`
      import { style } from '@mincho-js/css';
      const redColor = 'red';
      const red = style({ color: redColor });
      console.log(red);
    `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    it("inside block scope", () => {
      const { result, code } = babelTransform(`
      import { style } from '@mincho-js/css';
      {
        const red = style({ color: 'red' });
        console.log(red);
      }
    `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    it("already exported", () => {
      const { result, code } = babelTransform(`
      import { style } from '@mincho-js/css';

      export const red = style({ color: 'red' });
      console.log(red);
    `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    // doesn't work
    it("hoisting same variable name in different scope", () => {
      const { result, code } = babelTransform(`
      import { style } from '@mincho-js/css';

      const red = style({ color: 'red' });
      console.log(red);

      function SomeComponent() {
        const red = style({ color: 'blue' });
        console.log(red)
      }
    `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    // it("array pattern hoisting", () => {
    //   const { result, code } = babelTransform(`
    //   import { createTheme } from '@mincho-js/css';

    //   const [themeClass, vars] = createTheme({
    //     colors: {
    //       brand: 'red'
    //     }
    //   });
    //   console.log(themeClass, vars);
    // `);

    //   expect(result).toMatchSnapshot();
    //   expect(code).toMatchSnapshot();
    // });

    it("same binding in multiple declarations", () => {
      const { result, code } = babelTransform(`
      import { style } from '@mincho-js/css';

      const color = 'red';
      const foreground = style({ color });
      const background = style({ background: color });

      console.log(foreground, background)
    `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    it("binding ordering", () => {
      const { result, code } = babelTransform(`
      import { style } from '@mincho-js/css';

      const color = 'red';
      const red = style({ color });
      const longClass = \`abc \${red}\`;
      console.log(longClass)
    `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    it("nested bindings", () => {
      const { result, code } = babelTransform(`
      import { style } from '@mincho-js/css';

      const theme = { color: 'red' };
      const themeColor = theme.color;
      const color = themeColor;

      const red = style({ color });
      console.log(red)
    `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    // it("css variables", () => {
    //   const { result, code } = babelTransform(`
    //   import { style, createVar } from '@mincho-js/css';

    //   const colorVar = createVar();

    //   const red = style({
    //     color: 'red',
    //     vars: {
    //       [colorVar]: 'red'
    //     }
    //   });
    //   console.log(red)
    // `);

    //   expect(result).toMatchSnapshot();
    //   expect(code).toMatchSnapshot();
    // });

    it("global styles", () => {
      const { result, code } = babelTransform(`
      import { globalStyle } from '@mincho-js/css';

      globalStyle('html, body', {
        color: 'red',
      });
    `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    it("arrow function bindings", () => {
      {
        const { result, code } = babelTransform(`
        import { style } from '@mincho-js/css';

        const utility = { gap: (size) => ({ gap: size }) }
        const red = style({ ...utility.gap('10px') });
        console.log(red);
      `);

        expect(result).toMatchSnapshot();
        expect(code).toMatchSnapshot();
      }

      {
        const { result, code } = babelTransform(`
        import { style } from '@mincho-js/css';

        const getStyles = (color) => ({ color })
        const red = style({ ...getStyles('red') });
        console.log(red);
      `);

        expect(result).toMatchSnapshot();
        expect(code).toMatchSnapshot();
      }
    });

    it("function declaration bindings", () => {
      {
        const { result, code } = babelTransform(`
        import { style } from '@mincho-js/css';

        function getColor() { return 'red' }
        const red = style({ color: getColor() });
        console.log(red);
      `);

        expect(result).toMatchSnapshot();
        expect(code).toMatchSnapshot();
      }

      {
        const { result, code } = babelTransform(`
        import { style } from '@mincho-js/css';

        function getStyles(color) { return { color } }
        const red = style({ ...getStyles('red') });
        console.log(red);
      `);

        expect(result).toMatchSnapshot();
        expect(code).toMatchSnapshot();
      }
    });

    it("mincho-ignore doesn't extract expression", () => {
      const { result, code } = babelTransform(`
      import { style } from '@mincho-js/css';

      const red = /* mincho-ignore */ style({ color: "red" });
      console.log(red);
    `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });

    // it("react styled components get converted to recipe", () => {
    //   const { result, code } = babelTransform(`
    //   import { styled } from '@macaron-css/react';

    //   const Button = styled("button", {
    //     base: { color: 'red' }
    //   })
    //   console.log(Button)
    // `);

    //   expect(result).toMatchSnapshot();
    //   expect(code).toMatchSnapshot();
    // });

    // it("solid styled components get converted to recipe", () => {
    //   const { result, code } = babelTransform(`
    //   import { styled } from '@macaron-css/solid';

    //   const Button = styled("button", {
    //     base: { color: 'red' }
    //   })
    //   console.log(Button)
    // `);

    //   expect(result).toMatchSnapshot();
    //   expect(code).toMatchSnapshot();
    // });

    // it("leading comments of `styled` get passed to recipe", () => {
    //   const { result, code } = babelTransform(`
    //   import { styled } from '@macaron-css/solid';
    //   import {macaron$} from '@mincho-js/css';

    //   const fn = () => {
    //     const arr = [1,2]
    //     for (const _ of arr) {
    //       const Button = /* macaron-ignore */ styled("button", {
    //         base: { color: _ }
    //       })
    //     }
    //     return Button;
    //   }
    //   const test = macaron$(() => {
    //     console.log(fn())
    //     return "test"
    //   })
    //   console.log(test)
    // `);

    //   expect(result).toMatchSnapshot();
    //   expect(code).toMatchSnapshot();
    // });

    it("mincho-ignore on parent node", () => {
      const { result, code } = babelTransform(`
      import { globalStyle } from '@mincho-js/css';

      /* mincho-ignore */ globalStyle("html", { color: 'red' })
    `);

      expect(result).toMatchSnapshot();
      expect(code).toMatchSnapshot();
    });
  });
}
