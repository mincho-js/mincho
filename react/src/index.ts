import {
  PatternOptions,
  VariantGroups,
  VariantSelection,
} from '@vanilla-extract/recipes/dist/declarations/src/types';
import { PropsWithChildren, ReactElement, ReactNode } from 'react';

type Component<TProps = {}> = (props: TProps) => ReactNode;

type IntrinsicProps<TComponent> = TComponent extends keyof JSX.IntrinsicElements
  ? JSX.IntrinsicElements[TComponent]
  : any;

export function styled<TProps, Variants extends VariantGroups = {}>(
  component: Component<TProps>,
  options: PatternOptions<Variants>
): Component<PropsWithChildren<TProps & VariantSelection<Variants>>>;

export function styled<
  TProps,
  TComponent extends string | keyof JSX.IntrinsicElements,
  Variants extends VariantGroups = {}
>(
  component: TComponent,
  options: PatternOptions<Variants>
): Component<
  PropsWithChildren<IntrinsicProps<TComponent> & VariantSelection<Variants>>
>;

export function styled(
  component: any,
  options: any
): (props: any) => ReactElement {
  return function StyledComponent(props) {
    throw new Error(
      'This function should be stripped out at runtime. This error usually occurs if there is something wrong with the build configuration. If you think that the configuration is fine, then open an issue here `https://github.com/mokshit06/macaron/issues`'
    );
  };
}

export type StyleVariants<T extends (...args: any[]) => any> = Exclude<
  Parameters<T>[0]['variants'],
  undefined
>;
