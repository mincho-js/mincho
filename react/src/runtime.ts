import { useMemo, createElement } from 'react';

export function $$styled(
  Comp: any,
  styles: ((options?: any) => string) & {
    variants: Array<string>;
  }
) {
  function StyledComponent(props: any) {
    const allVariants = StyledComponent.variants;
    const [variants, others] = useMemo(() => {
      const [classes, others]: any[] = [{}, {}];

      for (const [key, value] of Object.entries(props)) {
        if (allVariants.includes(key)) {
          classes[key] = value;
        } else {
          others[key] = value;
        }
      }

      return [classes, others];
    }, [props]);
    const className = useMemo(() => {
      const classes = StyledComponent.classes(variants, props.className);
      return classes.join(' ');
    }, [variants, props.className]);

    if (typeof Comp === 'string') {
      return createElement(Comp, { ...others, className });
    }

    return createElement(Comp, { ...props, className });
  }

  StyledComponent.displayName = `Macaron(${Comp})`;
  StyledComponent.toString = () => StyledComponent.selector(null);
  StyledComponent.variants = [
    ...(styles.variants ?? []),
    ...(Comp.variants ?? []),
  ];
  StyledComponent.classes = (variants: any, merge?: string) => {
    const classes = new Set(
      classNames(styles({ ...variants }) + (merge ? ` ${merge}` : ''))
    );

    if (Comp.classes) {
      for (const c of Comp.classes({ ...variants }) as string[]) {
        classes.add(c);
      }
    }

    return Array.from(classes);
  };
  StyledComponent.selector = (variants: any) => {
    const classes = StyledComponent.classes(variants);

    // first element isn't empty
    if (classes.length > 0 && classes[0].length > 0) {
      return '.' + classes.join('.');
    }

    return classes.join('.');
  };

  return StyledComponent;
}

function classNames(className: string) {
  return className.split(' ');
}
