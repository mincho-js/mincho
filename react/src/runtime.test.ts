import { $$styled } from './runtime';
import { createElement } from 'react';
import { renderToString } from 'react-dom/server';
import { createRuntimeFn } from '../../core/src/create-runtime-fn';

function makeComponent() {
  return $$styled(
    'div',
    createRuntimeFn({
      defaultClassName: 'default',
      variantClassNames: {
        size: {
          sm: 'size_sm',
          md: 'size_md',
          lg: 'size_lg',
        },
        color: {
          light: 'color_light',
          dark: 'color_dark',
        },
      },
      compoundVariants: [],
      defaultVariants: {
        size: 'sm',
        color: 'light',
      },
    }) as any
  );
}

test('component has variants', () => {
  const Component = makeComponent();

  expect(Component.variants).toEqual(['size', 'color']);
});

test('component as selector', () => {
  const Component = makeComponent();

  expect(Component.toString()).toBe('.default');
  expect(`${Component}`).toBe('.default');

  expect(Component.selector({ size: 'md' })).toBe('.default.size_md');
  expect(Component.selector({ size: 'md', color: 'dark' })).toBe(
    '.default.size_md.color_dark'
  );
});

test('base component renders correctly', () => {
  const Component = makeComponent();

  expectRendersSnapshot(
    createElement(Component, { size: 'md', className: 'custom_extra_class' })
  );
});

test('inherit styled component', () => {
  const Component = makeComponent();
  const InheritedComponent = $$styled(
    Component,
    createRuntimeFn({
      defaultClassName: 'inherited',
      variantClassNames: {
        border: {
          true: 'border_true',
        },
      },
      compoundVariants: [],
      defaultVariants: {},
    }) as any
  );

  // expect(InheritedComponent.toString()).toBe('.inherited.default');
  // expect(`${InheritedComponent}`).toBe('.inherited.default');

  expect(InheritedComponent.selector({ size: 'md' })).toBe(
    '.inherited.default.size_md'
  );
  expect(InheritedComponent.selector({ size: 'md', color: 'dark' })).toBe(
    '.inherited.default.size_md.color_dark'
  );
  expect(InheritedComponent.selector({ border: true })).toBe(
    '.inherited.border_true.default'
  );

  expect(InheritedComponent.classes({})).toEqual(['inherited', 'default']);

  expectRendersSnapshot(createElement(InheritedComponent, {}));
  expectRendersSnapshot(
    createElement(InheritedComponent, {
      size: 'lg',
      className: 'custom_extra_cls',
    })
  );
});

function expectRendersSnapshot(component: any) {
  expect(renderToString(component)).toMatchSnapshot();
}
