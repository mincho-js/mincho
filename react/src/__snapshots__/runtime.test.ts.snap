// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`base component renders correctly 1`] = `"<div class=\\"default size_md color_light custom_extra_class\\"></div>"`;

exports[`inherit styled component 1`] = `"<div class=\\"default size_sm color_light inherited\\"></div>"`;

exports[`inherit styled component 2`] = `"<div class=\\"custom_extra_cls\\" class=\\"default size_lg color_light inherited\\"></div>"`;
